File: trees_and_graphs/maximum_depth_of_binary_tree.js
var maxDepthRecursive = function(root) {
    if (!root)
    {
        return 0;
    }

    const left = maxDepthRecursive(root.left);
    const right = maxDepthRecursive(root.right);
    return Math.max(left, right) + 1;
};

var maxDepthIterative = function(root) {
    if (!root)
    {
        return 0;
    }

    let stack = [[root, 1]];
    let ans = 0;

    while (stack.length)
    {
        const [node, depth] = stack.pop();
        ans = Math.max(ans, depth);

        if (node.left)
        {
            stack.push([node.left, depth + 1]);
        }

        if (node.right)
        {
            stack.push([node.right, depth + 1]);
        }
    }

    return ans;
};

File: trees_and_graphs/maximum_depth_of_binary_tree.py
import treenode

def maxDepthRecursive(root: treenode.TreeNode) -> int:
    if not root:
        return 0
    
    left = maxDepthRecursive(root.left)
    right = maxDepthRecursive(root.right)
    return max(left, right) + 1

def maxDepthIterative(root: treenode.TreeNode) -> int:
    if not root:
        return 0
    
    stack = [(root, 1)]
    ans = 0

    while stack:
        node, depth = stack.pop()
        ans = max(ans, depth)

        if node.left:
            stack.append((node.left, depth + 1))
        
        if node.right:
            stack.append((node.right, depth + 1))
    
    return ans